pkg/controlplane/control/manager.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	// create a k8s external service.
+	extName := eSpec.ExternalService
+	if extName != "" {
+		if net.IsIP(extName) {
+			extName += ".nip.io" // Convert IP to DNS address.
+		}
+
+		m.logger.Infof("Creating Kubernetes service %s of type ExternalName linked to %s.", eSpec.Service.Host, extName)
+
+		err := m.client.Create(
+			context.Background(),
+			&v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      eSpec.Service.Host,
+					Namespace: namespace,
+				},
+				Spec: v1.ServiceSpec{
+					Type:         v1.ServiceTypeExternalName,
+					ExternalName: extName,
+				},
+			})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeleteLegacyExport removes the possibility for ingress dataplane connections to access a given service.
+func (m *Manager) DeleteLegacyExport(namespace string, exportSpec *api.ExportSpec) error {
+	// Deleting a k8s external service.
+	if exportSpec.ExternalService != "" {
+		err := m.client.Delete(
+			context.Background(),
+			&v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      exportSpec.Service.Host,
+					Namespace: namespace,
+				},
+			})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// AddImport adds a listening socket for an imported remote service.
+func (m *Manager) AddImport(ctx context.Context, imp *v1alpha1.Import) error {
+	m.logger.Infof("Adding import '%s/%s'.", imp.Namespace, imp.Name)
+
+	// TODO: port manager should map ports to imports, and be able to detect conflicts
+	port, err := m.ports.Lease(imp.Spec.TargetPort)
+	if err != nil {
+		return fmt.Errorf("cannot generate listening port: %w", err)
+	}
+
+	imp.Spec.TargetPort = port
+
+	newService := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      imp.Name,
+			Namespace: imp.Namespace,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Protocol:   v1.ProtocolTCP,
+					Port:       int32(imp.Spec.Port),
+					TargetPort: intstr.FromInt32(int32(imp.Spec.TargetPort)),
+				},
+			},
+			Type:     v1.ServiceTypeClusterIP,
+			Selector: map[string]string{"app": dpapp.Name},
+		},
+	}
+
+	var oldService v1.Service
+	err = m.client.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      imp.Name,
+			Namespace: imp.Namespace,
+		},
+		&oldService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return m.client.Create(ctx, newService)
+		}
+
+		return err
+	}
+
+	if serviceChanged(&oldService, newService) {
+		return m.client.Update(ctx, newService)
+	}
+
+	return nil
+}
+
+// DeleteImport removes the listening socket of a previously imported service.
+func (m *Manager) DeleteImport(ctx context.Context, imp *v1alpha1.Import) error {
+	m.logger.Infof("Deleting import '%s/%s'.", imp.Namespace, imp.Name)
+	m.ports.Release(imp.Spec.TargetPort)
+	return m.client.Delete(
+		ctx,
+		&v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      imp.Name,
+				Namespace: imp.Namespace,
+			},
+		})
+}
+
+func serviceChanged(svc1, svc2 *v1.Service) bool {
```

Comment:
there isn't a better way to do it with deep copy?

Reply:
hmmm I searched google at the time and did not see anything that seems like a perfect solution.
Specifically came across this thread here:
https://github.com/kubernetes-sigs/kubebuilder/issues/592

But I'm open for new insights :)

Reply:
My fear is that the `ServiceSpec` returned from the API server can include some kind of auto-filled values which are not present on the `ServiceSpec` we build for creation.
@elevran your thoughts? Do you think it's safe to use `reflect.DeepEqual(oldService.Spec, newService.Spec)` ?

Reply:
based on the issue you posted, it isn't.
If the auto-filled fields are only updated once then a second call (e.g., status update) would be a nil-op.
However, your code would still need to check the relevant fields on the updates to see if it needs to reconcile or not, so might as well do it as you did here.